src/map/tileset.cpp

```diff
@@ -592,64 +610,69 @@ bool CTileset::isEquivalentTile(unsigned int tile1, unsigned int tile2) const
 	return mixedLookupTable[tile1] == mixedLookupTable[tile2];
 }
 
-int32_t CTileset::findTileIndexByTile(graphic_index tile) const
-{
-	tile_index index = 0;
-	for (const auto &checkTile : tiles) {
-		if (tile == checkTile.tile) {
-			return index;
-		}
-		index++;
-	}
-	return -1;
-}
-
 /**
-**  Get tile number.
+**  Get the indices of all tiles that belong to the same subslot as the tile with the given index.
+**  Each subslot contains tiles with identical type.
 **
-**  @param basic   Basic tile number
-**  @param random  Return random tile
-**  @param filler  Get a decorated tile.
+**  @param tileIndex	tile to find subslot neighbors for
 **
-**  @return        Tile index number.
+**  @return				vector of tile indices of tiles in the same subslot
 **
-**  @todo  FIXME: Solid tiles are here still hardcoded.
 */
-tile_index CTileset::getTileNumber(tile_index basic, bool random, bool filler) const
-{
-	tile_index tile = basic;
-	if (random) {
-		int n = 0;
-		for (int i = 0; i < 16; ++i) {
-			if (!tiles[tile + i].tile) {
-				if (!filler) {
-					break;
-				}
-			} else {
-				++n;
-			}
-		}
-		n = MyRand() % n;
-		int i = -1;
-		do {
-			while (++i < 16 && !tiles[tile + i].tile) {
+std::vector<tile_index> CTileset::getAllTilesOfTheSameKindAs(tile_index tileIndex) const
+{
+	auto calcRangeBoundary = [tileIndex, this](int dir) -> tile_index 
+	{
+		dir = dir < 0 ? -1 : 1;  // if dir < 0 then left boundary, right otherwise 
+		tile_index result = tileIndex;
+		if (getGraphicTileFor(tileIndex)) {
+			const int scopeWidth = dir < 0 ? tileIndex & 0xF
+										   : 0xF - (tileIndex & 0xF);
+			tile_index scope = 1;
+			while(scope <= scopeWidth) {
+				if (getGraphicTileFor(tileIndex + scope * dir)) result += dir;
+				else break;
+				scope++;
 			}
-		} while (i < 16 && n--);
-		if (i != 16) {
-			return tile + i;
 		}
+		return result;
+	};
+	const tile_index rangeFrom = calcRangeBoundary(-1);
+	const tile_index rangeTo = calcRangeBoundary(1);
+
+	std::vector<tile_index> result(rangeTo - rangeFrom + 1);
+	ranges::iota(result, rangeFrom);
+
+	return result;
+}
+
+bool CTileset::isTileRandomizable(tile_index tileIndex) const
+{
+	return CheckForUnseparatedSlot(getTile(tileIndex)) == false;
+}
+
+tile_index CTileset::getRandomTileOfTheSameKindAs(tile_index tileIndex) const
+{
+	if (!isTileRandomizable(tileIndex)) {
+		return tileIndex;
 	}
-	if (filler) {
-		int i = 0;
-		for (; i < 16 && tiles[tile + i].tile; ++i) {
-		}
-		for (; i < 16 && !tiles[tile + i].tile; ++i) {
-		}
-		if (i != 16) {
-			return tile + i;
+	auto tiles = getAllTilesOfTheSameKindAs(tileIndex);
+	if (tiles.empty()) {
+		return tileIndex;
+	}
+	return tiles[MyRand() % tiles.size()];	
+}
+
+int32_t CTileset::findTileIndexByTile(graphic_index tile) const
+{
+	tile_index index = 0;
+	for (const auto &checkTile : tiles) {
+		if (tile == checkTile.tile) {
+			return index;
 		}
+		index++;
 	}
-	return tile;
+	return -1;
```

Comment:
```suggestion
	if (auto it = ranges::find_if(tiles, [&](const auto &checkTile){ return tile == checkTile.tile; }; it != tiles.end()) {
		return std::distance(tiles.begin(), it);
	}
	return -1;
```

Reply:
Yeah, I agree.  I've been thinking about it myself. But I strongly dislike this idiom in terms of readability.

fixed:
https://github.com/Wargus/stratagus/pull/732/commits/fd85c8addb131e0e70eeade8a2a7353a793504cf